handlers/misttriggers/push_out_start.go

```diff
@@ -0,0 +1,82 @@
+package misttriggers
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/livepeer/catalyst-api/clients"
+	"github.com/livepeer/catalyst-api/errors"
+)
+
+// TriggerPushOutStart responds to PUSH_OUT_START trigger
+// This trigger is run right before an outgoing push is started. This trigger is stream-specific and must be blocking.
+// The payload for this trigger is multiple lines, each separated by a single newline character (without an ending newline), containing data:
+//
+// stream name (string)
+// push target URI (string)
+func (d *MistCallbackHandlersCollection) TriggerPushOutStart(w http.ResponseWriter, req *http.Request, payload []byte) {
+	lines := strings.Split(strings.TrimSuffix(string(payload), "\n"), "\n")
+	if len(lines) != 2 {
+		errors.WriteHTTPBadRequest(w, "Bad request payload", fmt.Errorf("unknown payload '%s'", string(payload)))
+		return
+	}
+	streamName := lines[0]
+	destination := lines[1]
+	var destinationToReturn string
+	switch streamNameToPipeline(streamName) {
+	case Recording:
+		destinationToReturn = d.RecordingPushOutStart(w, req, streamName, destination)
+	default:
+		destinationToReturn = destination
+	}
+	if _, err := w.Write([]byte(destinationToReturn)); err != nil {
+		log.Printf("TriggerPushOutStart failed to send rewritten url: %v", err)
+	}
+}
+
+func (d *MistCallbackHandlersCollection) RecordingPushOutStart(w http.ResponseWriter, req *http.Request, streamName, destination string) string {
+	event := &clients.RecordingEvent{
+		When:        "start",
+		Timestamp:   time.Now().UnixMilli(),
+		StreamId:    streamName,
+		RecordingId: uuid.New().String(),
+		Hostname:    req.Host,
+	}
+	pushUrl, err := url.Parse(destination)
+	if err != nil {
+		log.Printf("RecordingPushOutStart url.Parse %v", err)
+		return destination
+	}
+	if err = addUuidToUrl(pushUrl, event.RecordingId); err != nil {
+		log.Printf("RecordingPushOutStart addUuidToUrl() %v", err)
+		return destination
+	}
+	go clients.DefaultCallbackClient.SendRecordingEvent(event)
```

Comment:
I wonder if we should make these hooks blocking instead. As in if they fail we should not let the stream go through either. Only opening for consideration tho, not sure what i prefer.

Reply:
Good practice is to return as soon as possible so mistserver can continue without additional latency increase. 

Lets consider a watchdog process that kills bad streams separately from our happy-path.